filters/connection_state.go

```diff
@@ -0,0 +1,148 @@
+package filters
+
+import (
+	"net"
+	"net/http"
+)
+
+// ConnectionState holds data about a connection. No data is ever guaranteed to be present - the
+// getter methods, such as ConnectionState.Upstream, may return zero values for missing data.
+type ConnectionState struct {
+	// Values from this connection's first request.
+	originalURLScheme string
+	originalURLHost   string
+	originalHost      string
+
+	upstream, downstream net.Conn
+	upstreamAddr         string
+
+	requestNumber int
+
+	// Is this part of a MITM'd connection?
+	mitming bool
+
+	// Used by proxy.RequestAware connections.
+	requestAwareRequest  *http.Request
+	requestAwareUpstream net.Conn
```

Comment:
This is only a problem because the ConnectionState is strongly typed. If the ConnectionState where just a map (which is more like what context.Context was), the consumers of that map could handle the keys.

```go
type ConnectionState map[string]interface{}
```

If you want a nice interface, you could use a typedef in the `proxy` package like this:

```go
type ProxyConnectionState filters.ConnectionState

func (cs ProxyConnectionState) RequestAwareRequest() *http.Request {
    req, found := cs['req']
    if !found {
        return nil 
    }
    return req.(*http.Request)
}
```

Yes the type casting costs a little something, but I don't think we're using this stuff in a tight loop anywhere so I wouldn't be too concerned about it.

Reply:
That said, `filters` is only meaningful in the context of `proxy`, so the circularity you point out doesn't bother me all that much.

Reply:
I considered a map actually, but I think that creates the same kind of issues with subtle logic that we had with the context (for the same reason - it's just a bag-o-values).  With a struct, we can see which methods are called where.

> That said, filters is only meaningful in the context of proxy, so the circularity you point out doesn't bother me all that much.

Yeah, that fact that the circular bit is pretty hidden away by the API makes me more okay with it too (basically callers don't need to go back and forth between the packages, they can just implement the `RequestAware` and `ResponseAware` interfaces).